### PR TITLE
Web-Notifications —> Web

### PR DESCRIPTION
**Pull request by marianbitsov23**
### Stable Web Version
  - Notifications (Push + In-App)
  - Landing Page rework
  - Chat UX reworks
  - Bug Fixes
  - Now people that logged in externally, won't be able to edit their sensitive info

#### Reviews requested!

### CODE REVIEW

**Review by msimov02 — APPROVED** [review]
LGTM!

File: Web/tues_pairs/src/components/App/index.jsx [hunk 1]
```diff
@@ -22,8 +22,8 @@ import LandingPageBase from '../Pages/Landing/landing';
 const App = () => (
     <Router>
         <Navigation />
+        <Route exact path={ROUTES.LANDING} component={LandingPageBase} />
         <div className="main">
-            <Route exact path={ROUTES.LANDING} component={LandingPageBase} />
```
Comment (GGLozanov) [inline Web/tues_pairs/src/components/App/index.jsx:26]: What difference does this make?

File: Web/tues_pairs/src/components/Authentication/withAuthentication.jsx [hunk 2]
```diff
@@ -25,7 +25,33 @@ const withAuthentication = Component => {
...
+                                        deviceTokens.push(token);
+                                    }
+
+                                    this.props.firebase.db.collection("users").doc(authUser.uid).set({
+                                        deviceTokens: deviceTokens
+                                    }, { merge: true });
+
+                                    authUser.deviceTokens = deviceTokens;
+
+                                    this.setState({ authUser });
```
Comment (GGLozanov) [inline Web/tues_pairs/src/components/Authentication/withAuthentication.jsx]: Why not just use authUser.deviceTokens, or at least put it in a variable?

File: Web/tues_pairs/src/components/Firebase/firebase.jsx [hunk 3]
```diff
@@ -29,7 +30,22 @@ const config = {
...
+                deviceTokens.splice(deviceTokens.indexOf(token), 1);
+            }
+            this.db.collection("users").doc(currentUser.uid).set({
+                deviceTokens: deviceTokens
+            }, {merge: true});
+        }).then(() => {
+            this.auth.signOut();
+        })
+        .catch(error => {
+            console.error(error);
```
Comment (GGLozanov) [inline Web/tues_pairs/src/components/Firebase/firebase.jsx]: log.error()

File: Web/tues_pairs/src/components/Firebase/firebase.jsx [hunk 4]
```diff
@@ -92,6 +111,20 @@ const config = {
...
+        this.notifications().onSnapshot(snapshot => {
+            snapshot.forEach(doc => {
+                if(doc.data().userID === uid) {
+                    notifications.push({ ...doc.data(), nid: doc.id });
+                }
+            })
+        });
+
+        return notifications;
+    }
```
Comment (GGLozanov) [inline Web/tues_pairs/src/components/Firebase/firebase.jsx]: We can replace most of these filter functions that add overhead to the client-side application with a filter() query from Firestore (I believe JS also has it, like Flutter). I've made a note to replace my own queries with filter()/where(), so I suggest also taking those refactors into consideration here as well. Don't change for this PR, though.

File: Web/tues_pairs/src/components/Pages/Account/edit_personal_info.jsx [hunk 5]
```diff
@@ -236,6 +239,14 @@ class EditPersonalInfo extends Component{
 
         const hasImage = photoURL ? true : false;
 
+        let valid = true;
+
+        providerData.forEach(provider => {
+            if(provider.providerId !== 'password') {
+                valid = false;
+            }
+        });
```
Comment (GGLozanov) [inline Web/tues_pairs/src/components/Pages/Account/edit_personal_info.jsx]: Can you not extract this in ComponentDidMount() or one of the methods that trigger before the build state? I presume you only need this validation once for the state and not in each component render.

Also, I suppose you're checking whether the provider Id from the user's provider data matches the one for Firebase e-mail auth ("password") and setting the valid flag accordingly, correct?

File: Web/tues_pairs/src/constants/user_card.jsx [hunk 6]
```diff
@@ -30,11 +30,12 @@ const UserCard = props => {
                 {thisUser.isTeacher &&<Card.Subtitle data-testid="teacher">Teacher</Card.Subtitle>}
                 {!thisUser.isTeacher &&<Card.Subtitle data-testid="student">Student</Card.Subtitle>}
                 {!thisUser.isTeacher &&<Card.Text data-testid="gpa">GPA: {thisUser.GPA}</Card.Text>}
+                <Card.Text data-testid="email">{thisUser.email}</Card.Text>
                 {!thisUser.isTeacher &&<Card.Subtitle className="description" data-testid="student-idea">His idea</Card.Subtitle>}
                 {thisUser.isTeacher &&<Card.Subtitle className="description" data-testid="teacher-idea">Interested in</Card.Subtitle>}
                 {!thisUser.description &&<Card.Text data-testid="empty-description">Empty as the void inside our hearts!</Card.Text>}
                 <Card.Text data-testid="description">{thisUser.description}</Card.Text>
-                {show && <Card.Text>User has send you a match request</Card.Text>}
+                {show && <Card.Text>User has sent you a match request</Card.Text>}
```
Comment (GGLozanov) [inline Web/tues_pairs/src/constants/user_card.jsx:38]: Grammar is important! Good on you for fixing it.

**Review by GGLozanov — CHANGES_REQUESTED** [review]
You might want to check my comments (especially those in `edit_personal_info.jsx` and `withAuthentication.jsx`) and reply to them. If my observations there are incorrect, we can leave it at that (since the feature and code look good to me); otherwise, I'd suggest the small code improvements and optimisations referenced in said comments.

File: Web/tues_pairs/src/components/Pages/Account/edit_personal_info.jsx (hunk 5 above)
Comment (marianbitsov23) [reply Web/tues_pairs/src/components/Pages/Account/edit_personal_info.jsx]: Yes you are right, I moved it in the componentDidMount() method. Thanks for the request!

File: Web/tues_pairs/src/components/Firebase/firebase.jsx (hunk 3 above)
Comment (marianbitsov23) [reply Web/tues_pairs/src/components/Firebase/firebase.jsx]: Thanks, I will change it!

File: Web/tues_pairs/src/components/Authentication/withAuthentication.jsx (hunk 2 above)
Comment (marianbitsov23) [reply Web/tues_pairs/src/components/Authentication/withAuthentication.jsx]: Yes like that I will skip the use of the unneeded variable deviceTokens. I will simply use the authUser.deviceTokens to manipulate the array.

File: Web/tues_pairs/src/components/App/index.jsx (hunk 1 above)
Comment (marianbitsov23) [reply Web/tues_pairs/src/components/App/index.jsx:26]: I needed that for the design (css). I differentiate my main pages with this class="main". For the landing page I used background image and it's css was having some issues with this ```div className="main"```, so i extracted it outside, in order to have a responsive landing page.

File: Web/tues_pairs/src/components/Firebase/firebase.jsx (hunk 4 above)
Comment (marianbitsov23) [reply Web/tues_pairs/src/components/Firebase/firebase.jsx]: Okay I will do my research on firebase database queries https://firebase.google.com/docs/reference/js/firebase.database.Query
I will add a ToDo.

**Review by GGLozanov — APPROVED** [review]
Nice changes. Thanks. LGTM!

**Comment by marianbitsov23** [comment]
Thanks for the reviews! Changes approved!

**Pull request merged**